controllers/hvpa_controller.go

```diff
@@ -1077,7 +1104,7 @@ func (r *HvpaReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 
 	log.V(1).Info("Reconciling", "hvpa", instance.GetName())
 
-	r.manageCache(instance)
+	r.reconcileScaleStatus(instance)
```

Comment:
Reconciling HVPA's `scale` status before reconciling HPA, and before letting the scaling happen according to HPA could result in outdated value in the HVPA's scale status (both for target's selector and replicas).
Would it better to do this after `scaleIfRequired()` call?

Reply:
Target selector is not modified by reconciliation of HPA or VPA. The `replicas` in the status could be affected by both `reconcileHpa` and `scaleIfRequired`. But then we cannot expect the change to be immediate as `reconcileHpa` and `scaleIfRequired` will modify only the specs of HPA and VPA and there is no guarantee that the status would be updated by HPA and VPA within the same HVPA reconcile cycle.

Also, when HPA and VPA do update their status, HVPA reconciliation will be triggered and the HVPA status would be updated.

So, the order of reconciliation shouldn't matter.

One reason for keeping it where it is was that we could do reconcile `Scale` status and manage the cache at the same time. Not sure we should move the managing of cache after `scaleIfRequired`. WDYT?

Reply:
Another point is that when we have HPA pointing to the HVPA's `Scale` subresource, the status of the `targetRef` will certainly not be updated in the same reconciliation cycle.

Reply:
Consider the following case:
HVPA enters `Reconcile()` because HPA has new recommendation (Eg. scale up form 2 to 3). But the target is not scaled yet, because the `scaleIfRequierd()` is not called yet. The call to `reconcileScaleStatus()` updates the HVPA's `.status.replicas` to old number of replicas - 2. After HPA is reconciled and target is scaled, the number of replicas changes to 3.

In this case we would end up updating the HVPA's `.status.replicas` to 2, whereas number of actual replicas would be 3.

Also, should we consider calculating status of HVPA (once again...?) after all other steps that need to be taken by `Reconcile()`, as is done in any other controller (eg. KCM, MCM, etc.)?

Reply:
> After HPA is reconciled and target is scaled, the number of replicas changes to 3.

There is no guarantee that we will see this in the same HVPA reconcile cycle.

> Also, should we consider calculating status of HVPA (once again...?) after all other steps that need to be taken by Reconcile() complete, as is done in any other controller (eg. KCM, MCM, etc.)?

This is one option. Another could be a watch on the targetRef and re-enqueue HVPA. WDYT?

Reply:
>> After HPA is reconciled and **target is scaled**, the number of replicas changes to 3.

> There is no guarantee that we will see this in the same HVPA reconcile cycle.

This is keeping in mind that HVPA is responsible for scaling horizontally too which is where we are heading. Until then the scale subresource in HVPA has no meaning anyway, because we eventually want HPA controller to be the consumer of HVPA's scale subresource. If we continue with this implementation, we would need to revisit this anyway.

> Another could be a watch on the targetRef and re-enqueue HVPA.

Wouldn't it be an overkill, if this could be done just by using HPA's status which we anyway have after HPA's reconciliation. WDYT?